test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,341 @@
+#!/bin/bash
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2020 Intel Corporation. All rights reserved.
+
+set -e
+
+##
+## Case Name: check-volume-levels
+## Preconditions:
+##    topology is nocodec with loopack in PCM0P -> PCM0C
+##    aplay should work
+##    arecord should work
+##    PCM0C capture PGA supports -50 - +30 dB range
+##    PCM0C capture PGA supports mute switch
+## Description:
+##    Set volume and mute switch to various values, measure
+##    volume gain from the actual levels and compare to set gains.
+## Case step:
+##    1. Start aplay
+##    2a. Capture 1st wav file
+##    2b. Capture 2nd wav file
+##    2c. Capture 3rd wav file
+##    3. Measure volume gains
+## Expect result:
+##    command line check with $? without error
+##
+
+TESTDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
+
+# shellcheck source=case-lib/lib.sh
+source "$TESTDIR/case-lib/lib.sh"
+
+OPT_OPT_lst['t']='tplg'     OPT_DESC_lst['t']='tplg file, default value is env TPLG: $''TPLG'
+OPT_PARM_lst['t']=1         OPT_VALUE_lst['t']="$TPLG"
+
+func_opt_parse_option "$@"
+
+tplg=${OPT_VALUE_lst['t']}
+
+TPLGREADER="$TESTDIR"/tools/sof-tplgreader.py
+PCM_ID=0
+CAP_FORMAT="S16_LE"
+CAP_RATE="48000"
+VOL_P30DB=80
+VOL_P10DB=60
+VOL_0DB=50
+VOL_M10DB=40
+VOL_M20DB=30
+VOL_M30DB=20
+VOL_MAX=$VOL_P30DB
+VOL_MIN=1
+VOL_MUTE=0
+VOL_NOM=$VOL_0DB
+APLAY_WAV=$(mktemp --suffix=.wav)
+ARECORD_WAV1=$(mktemp --suffix=.wav)
+ARECORD_WAV2=$(mktemp --suffix=.wav)
+ARECORD_WAV3=$(mktemp --suffix=.wav)
+
+#
+# Main test procedure
+#
+
+main () {
+    # Preparations
+    do_preparations
+    generate_sine
+
+    # Start sine playback
+    aplay_helper "$APLAY_WAV" & aplayPID=$!
+    pause
+
+    # Do capture tests
+    run_test_1
+    run_test_2
+    run_test_3
+
+    # Stop sine playback and do cleanup
+    pause
+    dlogi "The test procedure is now complete. Killing the aplay process $aplayPID."
+    kill $aplayPID
+
+    # Measure, delete unnecessary wav files if passed
+    if measure_levels; then
+	dlogi "Deleting temporary files"
+	rm -f "$APLAY_WAV" "$ARECORD_WAV1" "$ARECORD_WAV2" "$ARECORD_WAV3"
+    fi
+}
```

Comment:
Since we have known issue https://github.com/thesofproject/sof/issues/3520 for nocodec SSP LBM, and this test case is designed to test the volume level tuning only, maybe better to stop the playback between capture iterations to avoid hitting the not related issue, some changes like this, I verified it can pass 10 iterations on my TGL nocodec.
```
diff --git a/test-case/check-volume-levels.sh b/test-case/check-volume-levels.sh
index b4ff6ed..40fe4c1 100755
--- a/test-case/check-volume-levels.sh
+++ b/test-case/check-volume-levels.sh
@@ -67,18 +67,28 @@ main () {
     generate_sine
 
     # Start sine playback
-    aplay_helper "$APLAY_WAV" & aplayPID=$!
+    aplay_helper "$APLAY_WAV" &
     pause
 
-    # Do capture tests
+    # Do capture test 1
     run_test_1
+
+    # Restart playback for capture test 2
+    killall aplay
+    pause
+    aplay_helper "$APLAY_WAV" &
     run_test_2
+
+    # Restart playback for capture test 3
+    killall aplay
+    pause
+    aplay_helper "$APLAY_WAV" &
     run_test_3
 
     # Stop sine playback and do cleanup
     pause
-    dlogi "The test procedure is now complete. Killing the aplay process $aplayPID."
-    kill $aplayPID
+    dlogi "The test procedure is now complete. Killing the aplay process."
+    killall aplay
 
     # Measure, delete unnecessary wav files if passed
     if measure_levels; then
```

Reply:
I used the continuous sine wave to lower the test execution time. It's especially useful if the test if electrical or acoustical. No need to wait two times more time before the playback is stable. Also there's more code "+" lines than "-" lines in your patch. I'm hesitating to change this just to avoid to trigger an issue. But no firm opinion. What do you think @marc-hb ?

Reply:
I understand it is this test that helped find issue thesofproject/sof#3520 so it sounds crazy to me to modify it so it *hides* issue 3520 before we even have another test to reproduce 3520.

While tests are generally focused on one thing they always "accidentally" test many other things and that's very good. The more bugs a test can find (in a resonable time) and the better test it is.

If we want to test volume levels in automation before 3520 is fixed then something like this should do it:
```shell
platform=$(sof-dump-status.py -p)
case $platform in
   byt|others)
   dloge "skipping this test until http://#3520 is fixed
   exit 2 # 2 means SKIP for Jenkins ;;
esac
```

Reply:
Yes, let's keep this, thanks.

Reply:
I don't prefer this skip, and this is so far manual run only. If the fails become a nuisance we can add this later.